go-controller/pkg/ovn/pod_selector_address_set.go

```diff
@@ -0,0 +1,639 @@
+		return
+	}
+	addrSetKey = getPodSelectorKey(podSelector, namespaceSelector, namespace)
+	err = oc.podSelectorAddressSets.DoWithLock(addrSetKey, func(key string) error {
+		psAddrSet, found := oc.podSelectorAddressSets.Load(key)
+		if !found {
+			psAddrSet = &PodSelectorAddressSet{
+				controllerName:    oc.controllerName,
+				key:               key,
+				backRefs:          map[string]bool{},
+				podSelector:       podSel,
+				namespaceSelector: nsSel,
+				namespace:         namespace,
+				addrSetDbIDs:      getPodSelectorAddrSetDbIDs(addrSetKey, oc.controllerName),
+			}
+			err = psAddrSet.init(oc)
+			// save object anyway for future use or cleanup
+			oc.podSelectorAddressSets.LoadOrStore(key, psAddrSet)
+			if err != nil {
+				psAddrSet.needsCleanup = true
+				return fmt.Errorf("failed to init pod selector address set %s: %v", addrSetKey, err)
+			}
+		}
+		if psAddrSet.needsCleanup {
+			cleanupErr := psAddrSet.destroy(oc)
+			if cleanupErr != nil {
+				return fmt.Errorf("failed to cleanup pod selector address set %s: %v", addrSetKey, err)
+			}
+			// psAddrSet.destroy will set psAddrSet.needsCleanup to false if no error was returned
+			// try to init again
+			err = psAddrSet.init(oc)
+			if err != nil {
+				psAddrSet.needsCleanup = true
+				return fmt.Errorf("failed to init pod selector address set %s after cleanup: %v", addrSetKey, err)
+			}
+		}
+		// psAddrSet is successfully inited, and doesn't need cleanup
+		psAddrSet.backRefs[backRef] = true
+		psAddrSetHashV4, psAddrSetHashV6 = psAddrSet.addressSet.GetASHashNames()
+		return nil
+	})
+	if err != nil {
+		err = fmt.Errorf("failed to GetPodSelectorAddressSet for caller %v: %w", backRef, err)
+		return
+	}
+	return
+}
+
+func (oc *DefaultNetworkController) DeletePodSelectorAddressSet(addrSetKey, backRef string) error {
+	return oc.podSelectorAddressSets.DoWithLock(addrSetKey, func(key string) error {
+		psAddrSet, found := oc.podSelectorAddressSets.Load(key)
+		if !found {
+			return nil
+		}
+		delete(psAddrSet.backRefs, backRef)
+		if len(psAddrSet.backRefs) == 0 {
+			err := psAddrSet.destroy(oc)
+			if err != nil {
+				// psAddrSet.destroy will set psAddrSet.needsCleanup to true in case of error,
+				// cleanup should be retried later
+				return fmt.Errorf("failed to destroy pod selector address set %s: %v", addrSetKey, err)
+			}
+			oc.podSelectorAddressSets.Delete(key)
+		}
+		return nil
+	})
+}
+
+func (psas *PodSelectorAddressSet) init(oc *DefaultNetworkController) error {
+	klog.Infof("Create shared address set for pod selector %s", psas.key)
+	psas.Lock()
+	pasLocked := true
+	defer func() {
+		if pasLocked {
+			psas.Unlock()
+		}
+	}()
+	if psas.addressSet == nil {
+		as, err := oc.addressSetFactory.NewAddressSet(psas.addrSetDbIDs, nil)
+		if err != nil {
+			return err
+		}
+		psas.addressSet = as
+	}
+	var err error
+	if psas.handler == nil {
+		// Unlock PodSelectorAddressSet before starting handlers to avoid deadlock,
```

Comment:
I dont see addPodSelectorHandler locking to set psAddrSet.handler = podHandler

Reply:
it doesn't, because the general logic is this 
```
// PodSelectorAddressSet methods are called from network policy handlers, and local handlers asynchronously.
// To make sure local handlers won't try to update deleted resources, this lock is used together with deleted field.
	sync.RWMutex
```
So for `PodSelectorAddressSet` the only object that need to be protected by lock is addressSet, and I actually don't even need to take the lock on init at all, only on delete.